src/Common/DNSResolver.cpp

```diff
@@ -246,38 +249,69 @@ String DNSResolver::getHostName()
 static const String & cacheElemToString(const String & str) { return str; }
 static String cacheElemToString(const Poco::Net::IPAddress & addr) { return addr.toString(); }
 
-template<typename UpdateF, typename ElemsT>
-bool DNSResolver::updateCacheImpl(UpdateF && update_func, ElemsT && elems, const String & log_msg)
+template <typename UpdateF, typename ElemsT>
+bool DNSResolver::updateCacheImpl(
+    UpdateF && update_func,
+    ElemsT && elems,
+    UInt32 max_consecutive_failures,
+    const String & notfound_log_msg,
+    const String & dropped_log_msg)
 {
     bool updated = false;
     String lost_elems;
-    for (const auto & elem : elems)
+    using iterators = typename std::remove_reference_t<decltype(elems)>::iterator;
+    std::vector<iterators> elements_to_drop;
+    for (auto it = elems.begin(); it != elems.end(); it++)
     {
         try
         {
-            updated |= (this->*update_func)(elem);
+            updated |= (this->*update_func)(it->first);
+            it->second = 0;
         }
-        catch (const Poco::Net::NetException &)
+        catch (const DB::Exception & e)
         {
+            if (e.code() != ErrorCodes::DNS_ERROR)
+            {
+                tryLogCurrentException(log, __PRETTY_FUNCTION__);
+                continue;
+            }
             ProfileEvents::increment(ProfileEvents::DNSError);
-
             if (!lost_elems.empty())
                 lost_elems += ", ";
-            lost_elems += cacheElemToString(elem);
+            lost_elems += cacheElemToString(it->first);
+            if (max_consecutive_failures)
+            {
+                it->second++;
+                if (it->second >= max_consecutive_failures)
+                    elements_to_drop.emplace_back(it);
+            }
         }
         catch (...)
         {
-            tryLogCurrentException(__PRETTY_FUNCTION__);
+            tryLogCurrentException(log, __PRETTY_FUNCTION__);
         }
     }
 
     if (!lost_elems.empty())
-        LOG_INFO(log, fmt::runtime(log_msg), lost_elems);
+        LOG_INFO(log, fmt::runtime(notfound_log_msg), lost_elems);
+    if (elements_to_drop.size())
```

Comment:
It makes sense to decrement failure counters and do not remove elements if `elements_to_drop.size() == elems.size()`. It may happen if DNS is not available for some reason and ClickHouse may stop working if DNS is not available and cache is cleared. However, trivial condition like `elements_to_drop.size() == elems.size()` will not work correctly, because names like `localhost` are resolved successfully even when DNS is not available. So I'm not sure what it the best solution, probably current one is ok.

Reply:
In the default scenario it's 15 seconds * 5 times so more than a minute without DNS to flush the cache. I think it's unlikely and even if it happens it means they will be deleted and added again (either because they are part of a cluster or a query) so I don't think it's a big deal

Reply:
Yeah, it's totally fine if it was a temporary short-time failure, elements will just be added again. But I mean that if DNS stops working at all then ClickHouse cluster will stop working too in 75 seconds. On the other hand, if you have a kind of Epic DNS Failure (I've seen some at Yandex) then broken ClickHouse is the least of your problems... So current solution looks ok for now.